src/toil/fileStore.py

```diff
@@ -1001,6 +1015,73 @@ def purgeRequired(self, jobReqs):
+                jobState = FileStore._JobState(cacheInfo.jobState[jobID])
+                logging.warning('Detected that job (%s) prematurely terminated.  Fixing the state '
+                                'of the cache.', jobState.jobName)
+                if not cleanup:
+                    logging.debug('The batch system is not in the process of shutting down.  Fixing'
+                                  ' the values in cache file so other jobs can run.')
+                    # Delete the old work directory if it still exists, to remove unwanted nlinks.
+                    # Do this only during the life of the program and dont' do it during the
+                    # batch system cleanup.  Leave that to the batch system cleanup code.
+                    if os.path.exists(jobState.jobDir):
+                        shutil.rmtree(jobState.jobDir)  # Ignore_errors?
+                    cacheInfo.sigmaJob -= jobState.jobReqs
+                # Run any deferred functions associated with the job
+                logging.debug('Running user-defined deferred functions.')
+                deferredFunctions = jobState.deferredFunctions
+                FileStore._runDeferredFunctions(deferredFunctions)
+                # Remove it from the cache state file
+                cacheInfo.jobState.pop(jobID)
+
+    def _registerDeferredFunction(self, func, *args, **kwargs):
+        """
+        This is the backend function to register a function that will be run after the completion of
+        the current job on the same node
+        :param function func: The function to be run after this job.
+        :param list args: The arguments to the function
+        :param dict kwargs: The keyword arguments to the function
+        """
+        with self._CacheState.open(self) as cacheInfo:
+            deferredFunction = DeferredFunction(func=func,
+                                                args=args,
+                                                kwargs=kwargs)
+            cacheInfo.jobState[self.hashedJobCommand]['deferredFunctions'].append(deferredFunction)
+            logging.error('Registered the deferred function "%s" to job "%s".', func.__name__,
+                          self.jobName)
+
+    @staticmethod
+    def _runDeferredFunctions(deferredFunctions):
+        """
+        This function will run all the registered deferred functions for the job.
+
+        :param list deferredFunctions: A list of DeferredFunctions to run
+        the cache
+        :return: None
+        """
+        for deferredFunction in deferredFunctions:
+            try:
+                logging.debug('Running deferred function "%s".', deferredFunction.func.__name__)
+                deferredFunction.func(*deferredFunction.args, **deferredFunction.kwargs)
+            except:
+                # This has to be generic because we don't know what is in the function
+                logging.warning('Deferred function "%s" failed.', deferredFunction.func.__name__)
```

Comment:
Consider also using logToMaster.

And makes sure stacktrace is included in worker log, such that someone can enable --logDebug and see that.